pallets/pallet-bonded-coins/src/try_state.rs

```diff
@@ -0,0 +1,83 @@
+use frame_support::traits::{
+	fungible::InspectHold,
+	fungibles::{metadata::Inspect as InspectMetadata, roles::Inspect as InspectRoles, Inspect},
+};
+use sp_runtime::{traits::Zero, TryRuntimeError};
+
+use crate::{types::PoolDetails, Config, FungiblesAssetIdOf, HoldReason, Pools};
+
+pub(crate) fn do_try_state<T: Config>() -> Result<(), TryRuntimeError> {
+	// checked currency ids. Each Currency should only be associated with one pool.
+	let mut checked_currency_ids = Vec::<FungiblesAssetIdOf<T>>::new();
+
+	Pools::<T>::iter().try_for_each(|(pool_id, pool_details)| {
+		let PoolDetails {
+			collateral_id,
+			deposit,
+			owner,
+			bonded_currencies,
+			state,
+			denomination,
+			..
+		} = pool_details;
+
+		let pool_account = pool_id.into();
+
+		// Collateral checks
+		let collateral_exists = T::Collaterals::asset_exists(collateral_id.clone());
+		assert!(collateral_exists);
+		let collateral_issuance = T::Collaterals::total_issuance(collateral_id);
+
+		// Deposit checks
+		let balance_on_hold_user =
+			T::DepositCurrency::balance_on_hold(&T::RuntimeHoldReason::from(HoldReason::Deposit), &owner);
+		assert!(balance_on_hold_user >= deposit);
+
+		// Bonded currencies checks
+		bonded_currencies
+			.iter()
+			.try_for_each(|currency_id| -> Result<(), TryRuntimeError> {
+				// check if currency is already associated with another pool
+				if checked_currency_ids.contains(currency_id) {
+					return Err(TryRuntimeError::Other(
+						"Currency is already associated with another pool",
+					));
+				}
```

Comment:
I'm curious, does the assert message show up at all in our tests or do we strictly need these errors for visibility? 
```suggestion
				assert!(!checked_currency_ids.contains(currency_id), "Currency is already associated with another pool");
```

Reply:
Updated in: [5e95f75](https://github.com/KILTprotocol/kilt-node/pull/825/commits/5e95f7549399b865a34ea9f6e045aaea12bf992d)